### ISSUE TITLE
WARNING : tensorflow:5 out of the last 5 calls to <function SpectralNormalization.normalize_weights at 0x7f24f9702160> triggered tf.function retracing.

### ISSUE THREAD

**Issue by MegaCreater** [tensorflow/addons#2575]
I am using `tfa.layers.SpectralNormalization` to normalise `tensorflow.keras.layers.Conv2D` weights. I had made the functional model api in which multiple conv2d (or SpectralNormalization) layer repeats, so i have used loop to do that. For example
`
inputs=tf.keras.layers.Input()
x=inputs
for x in range(5):
    x=SpectralNormalization(Conv2D())(x)
    x=ReLU(x)
`

Which give me multiple times, this warning (while calling with training=True) ->
`
WARNING:tensorflow:5 out of the last 5 calls to <function SpectralNormalization.normalize_weights at 0x7f24f9702160> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.
`
Q.1 Does this effect my model performance (GAN) ? 
Q.2 As it is due to 
`
@tf.function
def normalize_weights(self):
`
([build - source line106](https://github.com/tensorflow/addons/blob/v0.14.0/tensorflow_addons/layers/spectral_normalization.py#L106)). Can I make my own api and remove `tf.function` decorator ? will it work fine ? or add @tf.function(experimental_relax_shapes=True) (as suggested in warning) ?

**Comment by bhack** [comment]
You can try to use a x as tensor

See:

https://stackoverflow.com/questions/58972225/tensorflow2-warning-using-tffunction

**Comment by MegaCreater** [comment]
@bhack x is already input tensor. As it passed using input layer -> 

```
>>> input_layer=Input(shape=(128,128,3))
>>> type(input_layer)
<class 'tensorflow.python.keras.engine.keras_tensor.KerasTensor'>
 ```

and what I am passing that too is also tensor 
```
>>> sample_=tf.random.uniform([32,128,128,3],-1,1)
>>> outputs_=model_(sample_,training=True)
WARNING:tensorflow:5 out of the last 5 calls to <function SpectralNormalization.normalize_weights at 0x7f24f9702160> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details. ......
... (multiple warnings)
```

**Comment by bhack** [comment]
I meant in:

`for x in range(5):`

**Comment by MegaCreater** [comment]
@bhack ops! sorry for using same variable name. Its like ->
```
inputs=tf.keras.layers.Input()
x=inputs
for _ in range(5):
    x=SpectralNormalization(Conv2D())(x)
    x=ReLU(x)
```

**Comment by fsx950223** [comment]
I didn't meet the bug when I run following code:
```python
import tensorflow as tf
import tensorflow_addons as tfa

inputs=tf.keras.layers.Input([1, 320, 320, 3])
x=inputs
for i in range(50):
    x=tfa.layers.SpectralNormalization(tf.keras.layers.Conv2D(128, 3))(x)
    x=tf.keras.layers.ReLU()(x)

model=tf.keras.Model(inputs=[inputs], outputs=[x])
model.summary()
```

**Comment by bhack** [comment]
> I didn't meet the bug when I ran following code:
> ```python
> import tensorflow as tf
> import tensorflow_addons as tfa
> 
> inputs=tf.keras.layers.Input([1, 320, 320, 3])
> x=inputs
> for i in range(50):
>     x=tfa.layers.SpectralNormalization(tf.keras.layers.Conv2D(128, 3))(x)
>     x=tf.keras.layers.ReLU()(x)
> 
> model=tf.keras.Model(inputs=[inputs], outputs=[x])
> model.summary()
> ```
> Could you reproduce it on colab?

Yes, you cannot reproduce there cause the shape is the defined with the arg

**Comment by MegaCreater** [comment]
@bhack very true sir. This doesn't. Because class `tfa.layers.SpectralNormalization` doest call `@tf.function(def normalize_weights(self))` while model\layer build. It is called only when model is trained. Pass sample inputs like ->
```python3
import tensorflow as tf
import tensorflow_addons as tfa

inputs=tf.keras.layers.Input([320, 320, 3],batch_size=32)  # predefined input shape - > fixed
x=inputs
for i in range(50):
    x=tfa.layers.SpectralNormalization(tf.keras.layers.Conv2D(128, 3))(x)
    x=tf.keras.layers.ReLU()(x)

model_=tf.keras.Model(inputs=[inputs], outputs=[x])
#model_.summary()   # no need 
sample_inputs_=tf.random.truncated_normal([32, 320, 320, 3])  # 32 sample inputs
_=model_(sample_inputs_,training=True)
```
Look this colab file, which I made 4 you -> https://colab.research.google.com/drive/12WiRS_tS94X8H2KLrN22fTNzVIVjSo9r?usp=sharing

**Comment by bhack** [comment]
I think that the issue with you loop is that you are changing the input shape for every layer iteration. These are the first ten iteration:

```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(32, 320, 320, 3)]       0         
_________________________________________________________________
spectral_normalization (Spec (32, 318, 318, 128)       3712      
_________________________________________________________________
re_lu (ReLU)                 (32, 318, 318, 128)       0         
_________________________________________________________________
spectral_normalization_1 (Sp (32, 316, 316, 128)       147712    
_________________________________________________________________
re_lu_1 (ReLU)               (32, 316, 316, 128)       0         
_________________________________________________________________
spectral_normalization_2 (Sp (32, 314, 314, 128)       147712    
_________________________________________________________________
re_lu_2 (ReLU)               (32, 314, 314, 128)       0         
_________________________________________________________________
spectral_normalization_3 (Sp (32, 312, 312, 128)       147712    
_________________________________________________________________
re_lu_3 (ReLU)               (32, 312, 312, 128)       0         
_________________________________________________________________
spectral_normalization_4 (Sp (32, 310, 310, 128)       147712    
_________________________________________________________________
re_lu_4 (ReLU)               (32, 310, 310, 128)       0         
_________________________________________________________________
spectral_normalization_5 (Sp (32, 308, 308, 128)       147712    
_________________________________________________________________
re_lu_5 (ReLU)               (32, 308, 308, 128)       0         
_________________________________________________________________
spectral_normalization_6 (Sp (32, 306, 306, 128)       147712    
_________________________________________________________________
re_lu_6 (ReLU)               (32, 306, 306, 128)       0         
_________________________________________________________________
spectral_normalization_7 (Sp (32, 304, 304, 128)       147712    
_________________________________________________________________
re_lu_7 (ReLU)               (32, 304, 304, 128)       0         
_________________________________________________________________
spectral_normalization_8 (Sp (32, 302, 302, 128)       147712    
_________________________________________________________________
re_lu_8 (ReLU)               (32, 302, 302, 128)       0         
_________________________________________________________________
spectral_normalization_9 (Sp (32, 300, 300, 128)       147712    
_________________________________________________________________
re_lu_9 (ReLU)               (32, 300, 300, 128)       0         
=================================================================
```

**Comment by fsx950223** [comment]
The reason is that we create ```tf.function```  in a loop:
```python
import tensorflow as tf

def add():
    return 1

for i in range(15):
    tf.function(add)()
```
I removed internal ```tf.function``` in order to fix the bug.

**Comment by bhack** [comment]
> The reason is that we create ```tf.function```  in a loop:
> ```python
> import tensorflow as tf
> 
> def add():
>     return 1
> 
> for i in range(15):
>     tf.function(add)()
> ```
> I removed inner ```tf.function``` in order to fix the bug.
> 

Probably the issue Is in the 

Check the difference:

```
import tensorflow as tf

@tf.function
def add(i):
    print("Trace ",i)
    return i

for i in tf.range(15):
    print(add(i))

``

```
import tensorflow as tf

@tf.function
def add():
    print("Trace ")
    return 1

for i in range(15):
    print(add())
``

```
import tensorflow as tf

@tf.function
def add(i):
    print("Trace ",i)
    return i

for i in range(15):
    print(add(i))
```

**Comment by MegaCreater** [comment]
@fsx950223 thanks. But that is already know that this is due to **repeated call of @tf.function inside loop**  ([in spectral_normalization.py at L106](https://github.com/tensorflow/addons/blob/v0.14.0/tensorflow_addons/layers/spectral_normalization.py#L106)).
**I want to know removing this may/may effect my model performance?**

**Comment by MegaCreater** [comment]
@bhack thank a lot. But this should be removed from `class tfa.spectral_normalization.py` - [Line 106](https://github.com/tensorflow/addons/blob/v0.14.0/tensorflow_addons/layers/spectral_normalization.py#L106) will it effect ro slow down model training speed?

**Comment by bhack** [comment]
Also I've some general doubt about `normalize_weights` design with `self` only arg for the retracing monitoring as the 2d point in this comment:

https://github.com/tensorflow/tensorflow/issues/40631#issuecomment-649140316

**Comment by bhack** [comment]
> @bhack thank a lot. But this should be removed from `class tfa.spectral_normalization.py` - [Line 106](https://github.com/tensorflow/addons/blob/v0.14.0/tensorflow_addons/layers/spectral_normalization.py#L106). Will it effect ro slow down model training speed?

I don't know as we don't run real end2end benchmark in the CI so I cannot estimate the impact of this on some "real world" models. 
You can try to benchmark yourself with few layer.

The orginal review was at:
https://github.com/tensorflow/addons/pull/1244
